### ISSUE TITLE
NMS implementation

### ISSUE THREAD

**Issue by ZiyanZhu1994** [AlbertoSabater/Robust-and-efficient-post-processing-for-video-object-detection#12]
Hi, I am trying to apply REPP on custom videos. I noticed there are some overlapping bboxes for same object. I am wondering if I should apply NMS  on the outputs of REPP or there is an NMS module in REPP? 

Thanks in advance!

**Comment by AlbertoSabater** [comment]
Hi, NMS should be applied before REPP. [This ](https://github.com/AlbertoSabater/Robust-and-efficient-post-processing-for-video-object-detection/blob/d97feb2bc2d7f04c75cb939910f0855fff7ad21f/demos/YOLOv3/emodel.py#L545) is the sample of YOLOv3 in the paper. If you still find some overlapping you can modify the _iou_thr_ param of NMS.

Another cause of overlapping might be on the [prediction extraction from tubelets](https://github.com/AlbertoSabater/Robust-and-efficient-post-processing-for-video-object-detection/blob/master/REPP.py#L239). There, a different prediction is generated for each class with a score above _min_pred_score_, but the bbox coordinates will be the same.  You can try to augment this threshold or output just the detection generated for the most confident class.

**Comment by ZiyanZhu1994** [comment]
Thank you for your reply! I will modify the iou_thr of NMS.

I have another question that I am trying to finetune the embedding model. I am using another detector backbone instead of Yolo. From your paper, I learned that the embedding model are trained with triplet dataset. I am trying to find the training script for the embedding model, but only found the training script for logistic regression. Could you please point out the training script for the embedding model? 

Thank you!

**Comment by AlbertoSabater** [comment]
Unfortunately, I no longer have that code. However, the original training is very simple. The results reported in the article use the same triplets generated for the logistic regression as the dataset. Then for each training iteration, you sample a set of triplets (batch), augment the images and calculate the transformed patch coordinates for the objects, then extract the patches from the backbone and process them with the embedding model [RoIPooling, dense layer, normalization].  During training, I used the [triplet loss](https://github.com/AlbertoSabater/Robust-and-efficient-post-processing-for-video-object-detection/blob/d97feb2bc2d7f04c75cb939910f0855fff7ad21f/roi_nn.py#L130). 

Some improvements over this training (not implemented in the article) would be to change the triplet loss by a [semi-hard triplet loss](https://www.tensorflow.org/addons/api_docs/python/tfa/losses/triplet_semihard_loss) or the recent [Nt-Xent loss](https://github.com/margokhokhlova/NT_Xent_loss_tensorflow). In these cases, the triplets generated for the Logistic Regression would not longer be needed.
